swupd/rename.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// BSDIFFSIZE is smallest size of file to consider using bsdiff
+const BSDIFFSIZE = 200
+
+func renameDetection(manifest *Manifest, added []*File, removed []*File, c config) {
+	if len(added) == 0 || len(removed) == 0 {
+		return // nothing to rename
+	}
+	added = trimrenamed(added) // Make copies of input slices, tidy up whilst we are here
+	removed = trimrenamed(removed)
+	if err := fixupstatfields(removed, manifest, &c); err != nil {
+		panic(err)
+	}
+	// Handle pure renames first, don't need to worry about size. Should we skip zero size?
+	// just add call to trimsmall if so
+	// Sort by Hash. No need to have tiebreaker on name as the Manifest links by hash
+	// so 2 identical files being renamed, e.g. python2.6/foo and python3.7/bar being
+	// renamed to python2.7/foo and python3.7/bar have no concept which is the source for the
+	// python2.7/foo. This decision might need to change if the way renames are shown in
+	// the manifest changes.
+	sort.Slice(added, func(i, j int) bool {
+		return added[i].Hash < added[j].Hash
+	})
+	sort.Slice(removed, func(i, j int) bool {
+		return removed[i].Hash < removed[j].Hash
+	})
+	for ax, rx := 0, 0; ax < len(added) && rx < len(removed); {
+		af := added[ax]
+		rf := removed[rx]
+		switch {
+		case af.Hash < rf.Hash:
+			ax++
+		case af.Hash > rf.Hash:
+			rx++
+		default: // Equal hash, so link
+			linkrenamepair(af, rf)
+			ax++
```

Comment:
This means 'added index' I assume?

Reply:
yeah

Reply:
I said in one of my code reviews that using `i` and `j` wasn't very helpful as the two index values, as there was no connect in my mind between `i` and one of the added and removed slices, and therefore I needed to look very hard to see if the correct one was being incremented. Go programs traditionally use short variable names, and `i` is fine for a single index, but for this loop where we are stepping down two lists looking for matches I feel we need something better than `i` and `j`. Hence `ax` and `rx` for the added and removed index.

Reply:
Es ok, I figured it out pretty quickly. I just have PTSD from a different codebase using completely terrible names, so when I see 1-2 letter vars I look into it too hard :hammer:  Was mainly clarifying for myself/anyone else.